src/libraries/System.Runtime.Numerics/src/System/Numerics/BigNumber.cs

```diff
@@ -380,79 +382,254 @@ internal static BigInteger ParseBigInteger(ReadOnlySpan<char> value, NumberStyle
+        private static bool NumberToBigInteger(ref BigNumberBuffer number, out BigInteger result)
         {
-            int i = number.scale;
-            int cur = 0;
+            Span<int> stackBuffer = stackalloc int[BigInteger.StackallocUInt32Limit];
+            Span<int> currentBuffer = stackBuffer;
+            int currentBufferSize = 0;
+            int[]? arrayFromPool = null;
+
+            uint partialValue = 0;
+            int partialDigitCount = 0;
+            int totalDigitCount = 0;
+            int numberScale = number.scale;
+
+            const int MaxPartialDigits = 9;
+            const uint TenPowMaxPartial = 1000000000;
 
-            BigInteger ten = 10;
-            value = 0;
-            while (--i >= 0)
+            try
             {
-                value *= ten;
-                if (number.digits[cur] != '\0')
+                foreach (ReadOnlyMemory<char> digitsChunk in number.digits.GetChunks())
+                {
+                    if (!ProcessChunk(digitsChunk.Span, ref currentBuffer))
+                    {
+                        result = default;
+                        return false;
+                    }
+                }
+
+                if (partialDigitCount > 0)
+                    MultiplyAdd(ref currentBuffer, s_uint32PowersOfTen[partialDigitCount], partialValue);
+
+                int trailingZeroCount = numberScale - totalDigitCount;
+
+                while (trailingZeroCount >= MaxPartialDigits)
+                {
+                    MultiplyAdd(ref currentBuffer, TenPowMaxPartial, 0);
+                    trailingZeroCount -= MaxPartialDigits;
+                }
+
+                if (trailingZeroCount > 0)
+                    MultiplyAdd(ref currentBuffer, s_uint32PowersOfTen[trailingZeroCount], 0);
+
+                int sign;
+                uint[]? bits;
+
+                if (currentBufferSize == 0)
                 {
-                    value += number.digits[cur++] - '0';
+                    sign = 0;
+                    bits = null;
                 }
+                else if (currentBufferSize == 1 && (uint)currentBuffer[0] <= int.MaxValue)
+                {
+                    sign = number.sign ? -currentBuffer[0] : currentBuffer[0];
+                    bits = null;
+                }
+                else
+                {
+                    sign = number.sign ? -1 : 1;
+                    bits = MemoryMarshal.Cast<int, uint>(currentBuffer).Slice(0, currentBufferSize).ToArray();
+                }
+
+                result = new BigInteger(sign, bits);
+                return true;
             }
-            while (number.digits[cur] != '\0')
+            finally
             {
-                if (number.digits[cur++] != '0') return false; // Disallow non-zero trailing decimal places
+                if (arrayFromPool != null)
+                    ArrayPool<int>.Shared.Return(arrayFromPool);
```

Comment:
I don't see where `arrayFromPool` could be assigned in this try region?

Reply:
(Unless it's being implicitly captured below...)

Reply:
> (Unless it's being implicitly captured below...)

It is.